### DISCUSSION TITLE
use getInputProps with non-string values ? (number, boolean..)

### DISCUSSION THREAD

**Discussion by Pouet--** [edmundhung/conform#433] [Q&A]
I am trying to create a Form based on this zod schema : 
```tsx
const itemSchema = z.object({
  accessor: z.string(),
  width: z.number(),
  visible: z.boolean(),
});
```

and initialize it with some defaultValues. However, as you can see in [this sandbox](https://codesandbox.io/p/devbox/lvgkf8?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clsbsfxxk00062067p0hi30x4%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clsbsfxxk000220676lnmpwih%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clsbsfxxk00032067cnt44iow%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsbsfxxk000520679i8dhm3a%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clsbsfxxk000220676lnmpwih%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsbsfxxk00012067i7m2ydxn%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpublic%252Findex.html%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clsbsfxxk000220676lnmpwih%2522%252C%2522activeTabId%2522%253A%2522clsbsfxxk00012067i7m2ydxn%2522%257D%252C%2522clsbsfxxk000520679i8dhm3a%2522%253A%257B%2522id%2522%253A%2522clsbsfxxk000520679i8dhm3a%2522%252C%2522activeTabId%2522%253A%2522clsbskndm007p2067xjkdmv3i%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522clsbskndm007p2067xjkdmv3i%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%2522%257D%255D%257D%252C%2522clsbsfxxk00032067cnt44iow%2522%253A%257B%2522id%2522%253A%2522clsbsfxxk00032067cnt44iow%2522%252C%2522activeTabId%2522%253A%2522clsbskjcw004l2067an3heclo%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clsbskjcw004l2067an3heclo%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D), not all the fields are populated, and I have some Typescript errors with the use of getInputProps.

Moreover, when I update the fields individually, it does work with `string` values only, and fail with `number` and `boolean`, even if these values are accepted in the inputOptions.

Do I have to change the way of treating with these type of values ? Is it something related to my use of zod ? 
Thank you in advance for any recommendation :)

**Comment by austinzentzkippdc** [comment]
Conform does some work to convert common Javascript values to the corresponding HTML Input values. You can see some of the work [here](https://github.com/edmundhung/conform/blob/e0815d80a259c1e5bf657c2d04a0fe554d5a8476/packages/conform-react/helpers.ts#L265). This is a nice feature! You don't have to convert `undefined` or `null` to an empty string when you're dealing with a number type. But I think it's causing some confusion in your place.

[Here's](https://codesandbox.io/p/devbox/non-string-defaultvalue-forked-gtl5c9?file=%2Fsrc%2FApp.tsx%3A86%2C19&workspaceId=10676180-de78-460c-85ad-ad689ef0cec4) a simplified version of your sandbox that populates the values as expected. I made a couple of adjustments:
* I switched your input type from "boolean" to "checkbox. The type used in the `getInputProps` function must be an [HTML input type](https://www.w3schools.com/html/html_form_input_types.asp). 
* I standardized your number type so that the initial values were consistently numbers. This quieted typescript but I don't think this made a difference in the outputted code, though.

**Reply by Pouet--** [reply]
I did not notice the types used must be one of the HTML input types, thank you

**Comment by edmundhung** [comment]
Looks like there are some issues on the type inference. But all the default value is populated as expected. Conform handle all the value in the serialized format. This includes serializing the default value, like a boolean will become `"on"` if true or `""` if false. You can always serialize it yourself before passing it to the `defaultValue` option if you don't want the way the default serialization does.

With regards to the update, the value you are updating to will be serialized with the same rules. But I probably forgot to include this logic, so it end up broken in your example. I will fix it tonight.

The input type should be the HTML Input type as @aust1nz shared.

**Reply by edmundhung** [reply]
The issue with `form.update()` should be resolved in [v1.0.1](https://github.com/edmundhung/conform/releases/tag/v1.0.1).

**Reply by Pouet--** [reply]
It appears much clearer to me regarding the serialization, thank you 
(it works like a charm now)
